functional-programming-spreadsheet/index930.js

```diff
@@ -0,0 +1,122 @@
+const infixToFunction = {
+  "+": (x, y) => x + y,
+  "-": (x, y) => x - y,
+  "*": (x, y) => x * y,
+  "/": (x, y) => x / y
+};
+
+const infixEval = (str, regex) =>
+  str.replace(regex, (_, arg1, fn, arg2) =>
+    infixToFunction[fn](parseFloat(arg1), parseFloat(arg2))
+  );
+
+const highPrecedence = str => {
+  const regex = /([0-9.]+)([*\/])([0-9.]+)/;
+  const str2 = infixEval(str, regex);
+  return str === str2 ? str : highPrecedence(str2);
+};
+
+const isEven = num => num % 2 === 0;
+const sum = nums => nums.reduce((a, x) => a + x);
+const average = nums => sum(nums) / nums.length;
+
+const median = nums => {
+  const sorted = nums.slice().sort((x, y) => x - y);
+  const length = sorted.length;
+  const middle = sorted.length / 2 - 1;
+}; 
+
+const spreadsheetFunctions = {
+  "": x => x,
+  random: ([x, y]) => Math.floor(Math.random() * y + x),
+  increment: nums => nums.map(x => x + 1),
+  firsttwo: arr => arr.slice(0, 2),
+  lastttwo: arr => arr.slice(-2),
+  even: nums => nums.filter(isEven),
+  sum,
+  average,
+  has2: arr => arr.includes(2),
+  nodups: arr => arr.reduce((a, x) => a.includes(x) ? a : a.concat(x), []),
+  range: arr => range(...arr)
+};
+
+/*
+Finally, add a return statement, so that `median` returns the `average` of its two middle items if `length` `isEven`, and returns the middle item if it's not.
```

Comment:
This is quite complicated to explain in one step, so let's try to split it up a bit more. Still, even if it's split up, it would be fine to be more explicit about what learners should return.

The first step could be returning `isEven(length)`, and the second could be building out the ternary operator to return either `average([sorted[middle], sorted[middle + 1]])` or `sorted[middle + 0.5]`, along with the note about `middle` not being the actual middle of the array.